File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-no-legacy-expo-packages.ts
import { LEGACY_EXPO_PACKAGE_REPLACEMENTS } from "../../constants/react-native.js";
import { defineRule } from "../../utils/define-rule.js";
import { isTypeOnlyImport } from "../../utils/is-type-only-import.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const rnNoLegacyExpoPackages = defineRule({
  id: "rn-no-legacy-expo-packages",
  title: "Unmaintained legacy Expo package",
  tags: ["test-noise"],
  requires: ["react-native"],
  severity: "warn",
  recommendation:
    "Switch to the maintained replacement package so users are not stuck with unfixed bugs in deprecated Expo packages.",
  create: (context: RuleContext) => ({
    ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
      const source = node.source?.value;
      if (typeof source !== "string") return;
      if (isTypeOnlyImport(node)) return;

      for (const [packageName] of LEGACY_EXPO_PACKAGE_REPLACEMENTS) {
        if (source === packageName || source.startsWith(`${packageName}/`)) {
          context.report({
            node,
            message: `Your users are exposed to unfixed bugs when "${packageName}" is no longer maintained.`,
          });
          return;
        }
      }
    },
  }),
});

File: native/oxlint/rules/rn-no-legacy-expo-packages.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const LEGACY_EXPO_PACKAGES: [&str; 4] = [
    "expo-av",
    "expo-permissions",
    "expo-app-loading",
    "react-native-fast-image",
];

#[derive(Debug, Default, Clone)]
pub struct RnNoLegacyExpoPackages;

declare_oxc_lint!(
    /// Disallow unmaintained legacy Expo packages.
    RnNoLegacyExpoPackages,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow unmaintained legacy Expo packages.",
);

impl Rule for RnNoLegacyExpoPackages {
    fn should_run(&self, ctx: &crate::context::ContextHost) -> bool {
        !is_test_noise_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ImportDeclaration(import_declaration) = node.kind() else {
            return;
        };
        if is_type_only_import(import_declaration) {
            return;
        }
        let source = import_declaration.source.value.as_str();
        let Some(package_name) = LEGACY_EXPO_PACKAGES.iter().find(|package_name| {
            source == **package_name
                || source
                    .strip_prefix(**package_name)
                    .is_some_and(|suffix| suffix.starts_with('/'))
        }) else {
            return;
        };
        ctx.diagnostic(
            OxcDiagnostic::warn(format!(
                "Your users are exposed to unfixed bugs when \"{package_name}\" is no longer maintained."
            ))
            .with_label(import_declaration.span),
        );
    }
}
